runtime/interpreter/interpreter.go

```diff
@@ -2635,6 +2812,91 @@ func defineConverterFunctions(activation *VariableActivation) {
 	}
 }
 
+type runtimeTypeConstructor struct {
+	name      string
+	converter *HostFunctionValue
+}
+
+// Constructor functions are stateless functions. Hence they can be re-used across interpreters.
+//
+var runtimeTypeConstructors = []runtimeTypeConstructor{
+	{
+		name: "OptionalType",
+		converter: NewHostFunctionValue(
+			func(invocation Invocation) Value {
+				return TypeValue{
+					Type: OptionalStaticType{
+						Type: invocation.Arguments[0].(TypeValue).Type,
+					}}
+
+			},
+			sema.OptionalTypeFunctionType,
+		),
+	},
+	{
+		name: "VariableSizedArrayType",
+		converter: NewHostFunctionValue(
+			func(invocation Invocation) Value {
+				return TypeValue{
+					Type: VariableSizedStaticType{
+						Type: invocation.Arguments[0].(TypeValue).Type,
+					}}
+			},
+			sema.VariableSizedArrayTypeFunctionType,
+		),
+	},
+	{
+		name: "ConstantSizedArrayType",
+		converter: NewHostFunctionValue(
+			func(invocation Invocation) Value {
+				return TypeValue{
+					Type: ConstantSizedStaticType{
+						Type: invocation.Arguments[0].(TypeValue).Type,
+						Size: int64(invocation.Arguments[1].(IntValue).ToInt()),
+					}}
+			},
+			sema.ConstantSizedArrayTypeFunctionType,
+		),
+	},
+	{
+		name: "ReferenceType",
+		converter: NewHostFunctionValue(
+			func(invocation Invocation) Value {
+				return TypeValue{
+					Type: ReferenceStaticType{
+						Authorized: bool(invocation.Arguments[0].(BoolValue)),
+						Type:       invocation.Arguments[1].(TypeValue).Type,
+					}}
+			},
+			sema.ReferenceTypeFunctionType,
+		),
+	},
+	{
+		name: "CapabilityType",
+		converter: NewHostFunctionValue(
+			func(invocation Invocation) Value {
+				switch ty := invocation.Arguments[0].(TypeValue).Type.(type) {
+				case ReferenceStaticType:
+					return TypeValue{
+						Type: CapabilityStaticType{
+							BorrowType: ty,
+						}}
+				// Capabilities must hold references
+				default:
+					return NilValue{}
+				}
```

Comment:
Declare input, bail-out early, and most importantly, return an optional:

```suggestion
				ty := invocation.Arguments[0].(TypeValue).Type
				
				// Capabilities must hold references
				_, ok := ty.(ReferenceStaticType)
				if !ok {
					return NilValue{}
				}
				
				return NewSomeValueOwningNonCopying(
					TypeValue{
						Type: CapabilityStaticType{
							BorrowType: ty,
						},
					},
				)
```